R/source.R

```diff
@@ -282,22 +270,63 @@ invoke_cargo <- function(toolchain, specific_target, dir, profile,
     # If RTOOLS40_HOME is properly set, this will have no real effect
     withr::local_envvar(RTOOLS40_HOME = rtools_home)
   }
-  status <- system2(
+
+  message_buffer <- character(0)
+  env <- rlang::current_env()
+
+  tty_has_colors <- isTRUE(cli::num_ansi_colors() > 1L)
+
+  compilation_result <- processx::run(
     command = "cargo",
     args = c(
       sprintf("+%s", toolchain),
       "build",
       "--lib",
-      if (!is.null(specific_target)) sprintf("--target %s", specific_target) else NULL,
-      sprintf("--manifest-path %s", file.path(dir, "Cargo.toml")),
-      sprintf("--target-dir %s", file.path(dir, "target")),
-      if (profile == "release") "--release" else NULL
+      if (!is.null(specific_target)) {
+        sprintf("--target=%s", specific_target)
+      } else {
+        NULL
+      },
+      sprintf("--manifest-path=%s", file.path(dir, "Cargo.toml")),
+      sprintf("--target-dir=%s", file.path(dir, "target")),
+      if (profile == "release") "--release" else NULL,
+      "--message-format=json-diagnostic-rendered-ansi",
+      ifelse(
+        tty_has_colors,
+        "--color=always",
+        "--color=never"
+      )
     ),
-    stdout = stdout,
-    stderr = stderr
+    echo_cmd = FALSE,
+    windows_verbatim_args = FALSE,
+    stderr = ifelse(isTRUE(quiet), "|", ""),
+    stdout = "|",
+    error_on_status = FALSE,
+    stdout_line_callback = function(line, ...) {
+      assign("message_buffer", c(message_buffer, line), envir = env)
+    }
   )
-  if (status != 0L) {
-    ui_throw("Rust code could not be compiled successfully. Aborting.")
+
+  if (!isTRUE(compilation_result$status == 0)) {
+    errors <-
+      message_buffer %>%
+      purrr::map_chr(stringi::stri_trim) %>%
+      purrr::map(jsonlite::parse_json) %>%
+      purrr::keep(
+        ~.x$reason == "compiler-message" && .x$message$level == "error"
+      )
```

Comment:
Here I rely on the fact that `cargo` output is configured to produce JSONs. Should we have an extra `tryCatch` to handle a situation when the output is garbage and can't be parsed as JSON?